### PR TITLE
hypervisor, vmm: Abstract the interfaces to start/stop dirty log to support both KVM and MSHV

### PR DESCRIPTION
**Pull request by likebreath**
Following KVM interfaces, the `hypervisor` crate now provides interfaces
to start/stop the dirty pages logging on a per region basis, and asks
its users (e.g. the `vmm` crate) to iterate over the regions that needs
dirty pages log. MSHV only has a global control to start/stop dirty
pages log on all regions at once.

This patch refactors related APIs from the `hypervisor` crate to provide
a global control to start/stop dirty pages log (following MSHV's
behaviors), and keeps tracking the regions need dirty pages log for
KVM. It avoids leaking hypervisor-specific behaviors out of the
`hypervisor` crate.

This use the above changes, we also need to explicitly tell the hypervisor (KVM)
whether a region needs dirty-pages-log when it is created. This is done through
reverting commit  f063346.

Signed-off-by: Bo Chen <chen.bo@intel.com>

### CODE REVIEW

File: hypervisor/src/kvm/mod.rs
Comment (russell-islam) [inline hypervisor/src/kvm/mod.rs:269]: This seems wrong to me. This function is used to make the MemoryRegion(both KVM and MSHV)  parameter , this is not the setting the memory regions. We should populate the slots in the create_user_memory_region call and remove the slot in the remove_user_memory_region function.

**Review by russell-islam — CHANGES_REQUESTED** [review]

File: hypervisor/src/kvm/mod.rs
Comment (likebreath) [reply hypervisor/src/kvm/mod.rs:269]: Good point. Will update the PR shortly. 

If there is anything else missing (particularly for you to integrate with MSHV), please let me know. Thank you.

Comment (russell-islam) [reply hypervisor/src/kvm/mod.rs:269]: Otherwise seems good to me.

Comment (likebreath) [reply hypervisor/src/kvm/mod.rs]: Cool. Updated. PTAL.

**Review by russell-islam — APPROVED** [review]

**Comment by russell-islam** [comment]
@likebreath Can we move forward with this PR so that my MSHV PR can move as well.

**Comment by likebreath** [comment]
> Can we move forward with this PR so that my MSHV PR can move as well.

Yes, that's the plan. This PR needs another approval from @rbradford or @sboeuf before we can land it.

File: hypervisor/src/kvm/mod.rs
Comment (russell-islam) [inline hypervisor/src/kvm/mod.rs:282]: When this is true? Does the caller ever set this flag?

**Review by russell-islam — CHANGES_REQUESTED** [review]

File: hypervisor/src/kvm/mod.rs
Comment (likebreath) [reply hypervisor/src/kvm/mod.rs:282]: Yes, it is  expected to be set by the caller/user, just like all other members of `MemoryRegion`. 

For the current usage from the `vmm` crate, this flag when calling into `make_user_memory_region` with `log_dirty_pages = true` (check the second commit for details).

**Review by russell-islam — APPROVED** [review]

**Review by rbradford — APPROVED** [review]

**Pull request merged**
